matching/README.md

```diff
@@ -0,0 +1,991 @@
+except it switches over the runtime length of the list and makes sure to add
+bindings for each list element being matched on to the substitution. Note that
+in the case of the default, we know that the list must be greater in length
+than any list pattern in the column being decomposed, so we only must decompose
+maxHead elements from the front of the list and maxTail elements from the end,
+since we can guarantee any elements in between will never be referenced by any
+submatrix. This is the reason why we treat the default operation as equivalent
+to specializing on `maxHead + maxTail` elements.
+
+Finally, we must define `foldRanges`, which handles the R_i sections of a
+clause C. This ensures that a binding for the L variable in list patterns with
+variables exists by the time a rule is applied. The reason we defer this until
+the end of the decision tree is that different rows may need to construct the
+value of L differently even given the same runtime length of the list. Thus,
+we cannot simply do it as part of the specialization process, or we will
+disrupt sharing for the decision tree.
+
+```
+def foldRanges([(Occurrence, Sort, Int, Int)] R, DT dt): DT
+  case R of
+    [] -> dt
+    (l : o, s, h, t) : tail ->
+      val f = the function that removes h elements from the head of a list and t elements from its tail
+      foldRanges(tail, Function(l : o, f, [(o, s)], s, dt))
+```
+
+### Canonicalized Patterns
+
+Thus far we have not really given thought to how we define variables.
+
+In most contexts, a variable is simply a name:
+
+```
+syntax Variable ::= String
+```
+
+However, there are a few contexts where using named variables is problematic.
+In particular, in the `Constructor`, `Occurrence`, and `DT` sorts, we actually
+need to be able to reference a variable by the occurrence that bound it. Thus,
+in these contexts, we actually define a variable as follows:
+
+```
+syntax Variable ::= Occurrence
+```
+
+We thus have two notions of patterns: patterns whose variables are names, and
+patterns whose variables are occurrences. As a result, we need some way of
+converting between these representations.
+
+As it happens, the clause list of a matrix provides this information in the
+form of the substitution S. We thus say that to *canonicalize* a pattern p
+with respect to row i of a Matrix m is to substitute all named variables in p
+with their bindings in S_i. We say that a pattern is *bound* if all variables
+in p have a binding in S_i.
+
+### Map Patterns
+
+K does not implement full AC matching on maps or sets. Instead, it attempts to
+order the matching process in such a manner that wherever possible, the
+substitution for all variables in the keys of map patterns are bound before
+matching on the map itself. This allows us to, in the vast majority of cases,
+compile matching on maps to map membership and lookup checks, which are vastly
+more efficient than AC matching.
```

Comment:
For a technical document this is fine here, but for a paper we'll probably want to shift it closer to the start to understand the motivation for extending Maranget with Lists / Sets / Maps.

Reply:
By the way - do you have an intuitive example of what happens if you _don't_ specialise your implementation for collections, and instead just implement them as algebraic types (i.e. use their "as-if" version)? You'd end up with horrible-to-write code, and wouldn't be able to abstract over collections of different size, but what would the difference in the generated trees look like?

Reply:
That makes sense - thanks